src/celeritas/track/detail/TrackSortUtils.cu

```diff
@@ -59,14 +60,39 @@ void partition_impl(TrackSlots const& track_slots, F&& func, StreamId stream_id)
 
 //---------------------------------------------------------------------------//
 
-template<class F>
-void sort_impl(TrackSlots const& track_slots, F&& func, StreamId stream_id)
+__global__ void
+reorder_actions_kernel(ObserverPtr<TrackSlotId::size_type const> track_slots,
+                       ObserverPtr<ActionId const> actions,
+                       ObserverPtr<ActionId::size_type> out_actions,
+                       size_type size)
 {
-    auto start = device_pointer_cast(track_slots.data());
-    thrust::sort(thrust::device.on(celeritas::device().stream(stream_id).get()),
-                 start,
-                 start + track_slots.size(),
-                 std::forward<F>(func));
+    if (ThreadId tid = celeritas::KernelParamCalculator::thread_id();
+        tid < size)
+    {
+        out_actions.get()[tid.get()]
+            = actions.get()[track_slots.get()[tid.get()]].get();
```

Comment:
How many get can we get on one statement?

Reply:
There is an open issue about improving the nomenclature 😅